src/Components/Web.JS/src/Services/NavigationManager.ts

```diff
@@ -116,20 +178,111 @@ function performInternalNavigation(absoluteInternalHref: string, interceptedLink
   } else {
-    history.replaceState(state, /* ignored title */ '', absoluteInternalHref);
+    history.replaceState({
+      userState: state,
+      index: currentHistoryIndex,
+    }, /* ignored title */ '', absoluteInternalHref);
   }
 
   notifyLocationChanged(interceptedLink);
 }
 
+function navigateHistoryWithoutPopStateCallback(delta: number): Promise<void> {
+  return new Promise(resolve => {
+    const oldPopStateCallback = popStateCallback;
+
+    popStateCallback = () => {
+      popStateCallback = oldPopStateCallback;
+      resolve();
+    };
+
+    history.go(delta);
+  });
+}
+
+let currentNotifyLocationChangingPromise: Promise<boolean> | null = null;
+
+function ignorePendingNavigations() {
+  // This method ensures that we don't perform multiple overlapping navigations
+  // that bypassed the .NET cancellation logic due to network latency.
+
+  // It's possible for a JS -> .NET 'location changing' message to be in flight
+  // while a .NET -> JS response for a previous 'location changing' event is on its way.
+  // Since the .NET side only processes one navigation at a time in this case, the navigations
+  // aren't considered "overlapping". It's up to the JS side to recognize this case so we don't
+  // navigate in ways the user did not intend (e.g., we don't want navigations to previous entries
+  // in the browser tab's history to accumulate).
+  currentNotifyLocationChangingPromise = null;
+}
+
+async function notifyLocationChanging(uri: string, intercepted: boolean): Promise<boolean> {
+  if (!notifyLocationChangingCallback) {
+    return false;
+  }
+
+  const notifyLocationChangingPromise = notifyLocationChangingCallback(uri, intercepted);
+  currentNotifyLocationChangingPromise = notifyLocationChangingPromise;
+  const navigationCanceled = await notifyLocationChangingPromise;
+
+  return navigationCanceled || currentNotifyLocationChangingPromise !== notifyLocationChangingPromise;
+}
+
+async function onBrowserInitiatedPopState(state: PopStateEvent) {
+  ignorePendingNavigations();
+
+  const index = state.state?.index ?? 0;
+  const delta = index - currentHistoryIndex;
+  const uri = location.href;
+
+  if (hasLocationChangingEventListeners) {
+    // Temporarily revert the navigation until we confirm if the navigation should continue.
+    await navigateHistoryWithoutPopStateCallback(-delta);
+
+    // Invoke the "location changing" event first to prioritize programmatic cancellation.
+    let navigationCanceled = await notifyLocationChanging(uri, false);
+    if (navigationCanceled) {
+      return;
+    }
+
+    // Allow the user to cancel the navigation, if necessary.
+    navigationCanceled = displayNavigationPromptIfEnabled();
+    if (navigationCanceled) {
+      return;
+    }
+
+    await navigateHistoryWithoutPopStateCallback(delta);
+  } else {
+    // Allow the user to cancel the navigation, if necessary.
+    const navigationCanceled = displayNavigationPromptIfEnabled();
```

Comment:
I'm still wondering if this has to be baked into NavigationManager itself. Would it be practical for the NavigationLock component to deal with the `beforeunload` event handler stuff completely separately? That is, everything to do with external navigations.

Or is there some necessity for NavigationManager to have a concept of external navigations?

Reply:
I've pulled out the `<NavigationLock />`-specific code into its own `.ts` file.